packages/hash/api/src/model/entityType.util.ts

```diff
@@ -0,0 +1,288 @@
+ * */
+export const generateSchema$id = (
+  accountId: string,
+  entityTypeId: string,
+  relative: boolean = false,
+) => {
+  return `${relative ? "" : FRONTEND_URL}/${accountId}/types/${entityTypeId}`;
+};
+
+/**
+ * Given a Schema$id, generate an appropriate $ref to put into allOf field on JSON schema.
+ * This can be used to inherit from other schemas.
+ * */
+export const createSchema$idRef = (schema$id: string) => {
+  return JSON.stringify([{ $ref: schema$id }]);
+};
+
+// The following types are internally used to validate JSON schemas.
+type Property = {
+  name: string;
+  type: string;
+  format?: string;
+  description?: string;
+  otherFields: Partial<JSONSchema>;
+};
+
+// If any property constraint has 'min' or 'max' in their name, it is a numeric constraint
+// JSON Schema 4 would allow booleans for 'exclusiveMinimum' and 'exclusiveMaximum', though.
+const CONSTRAINT_RE = /.*(min|max).*/i;
+
+/**
+ * When a 'minimim' constraint is defined twice on the same property (across multiple schemas),
+ * the constraint should be 'maximized' such that the minimum is changed to the most permissive
+ * value.
+ */
+const maximizeConstraint = Math.max;
+
+/**
+ * When a 'maximum' constraint is defined twice on the same property (across multiple schemas),
+ * the constraint should be 'minimized' such that the maximum is changed to the least permissive
+ * value.
+ */
+const minimizeConstraint = Math.min;
+
+/**
+ * All JSON Schema numeric constraints that we support validation for.
+ */
+type Constraint =
+  | "exclusiveMaximum"
+  | "exclusiveMinimum"
+  | "maximum"
+  | "minimum"
+  | "maxItems"
+  | "minItems"
+  | "maxLength"
+  | "minLength"
+  | "maxProperties"
+  | "minProperties";
+
+/**
+ * Mapping from constraint names to what kind of permissiveness that is required
+ */
+const propertyConstraintMerging: Record<
+  Constraint | string,
+  (...values: number[]) => number
+> = {
+  exclusiveMaximum: minimizeConstraint,
+  exclusiveMinimum: maximizeConstraint,
+  maximum: minimizeConstraint,
+  minimum: maximizeConstraint,
+  maxItems: minimizeConstraint,
+  minItems: maximizeConstraint,
+  maxLength: minimizeConstraint,
+  minLength: maximizeConstraint,
+  maxProperties: minimizeConstraint,
+  minProperties: maximizeConstraint,
+} as const;
+
+/**
+ * Traverse properties of a schema and convert to a {@link Property}
+ *
+ * @param schema JSON Schema that is to be traversed
+ */
+function extractProperties(schema: JSONSchema): Property[] {
+  const properties: Property[] = [];
+
+  for (const [field, value] of Object.entries(schema.properties ?? {})) {
+    if (typeof value === "object") {
+      const { type, format, description, ...otherFields } = value;
+      if (typeof type === "string") {
+        properties.push({
+          name: field,
+          type,
+          format,
+          description,
+          otherFields,
+        });
+      }
+    }
+  }
+
+  return properties;
+}
```

Comment:
You're likely already aware @thehabbos007, but having read through this file made me realise we are only really validating property definitions made at the root level of a JSON Schema's `properties` field.

This means we are not accounting for definitions made in other places, for instance arrays:
```JSON
"someProperty": {
  "type": "array",
  "items": {
    "type": "number",
    "min": 0,
    "max": 10,
  }
}
```
Or nested objects:
```JSON
"someProperty": {
  "type": "object",
  "properties": {
    "nestedProperty": {
      "type": "number",
      "min": 0,
      "max": 10,
    }
  }
}
```
Or [definitions made in sub-schemas](https://json-schema.org/understanding-json-schema/structuring.html), and likely multiple other places definitions could occur that I haven't thought of just now. 

I'm not sure to what extent we want to support these JSON-Schema definitions (perhaps we should have an internal discussion about this, cc @CiaranMn), but nonetheless I feel compelled to point out that we're currently not accounting for them in this file.

Reply:
Yes, correct. It's definitely something we'll incrementally add as (or if) we need it :)

Reply:
We will definitely support array fields, and thus should validate constraints in them.

There's not yet any provision for nested objects in [the editor design](https://www.figma.com/file/gydVGka9FjNEg9E2STwhi2/HASH-Editor?node-id=807%3A65588) so it's unclear if we will need to support/validate properties in them (whether they're defined in a nested object in the schema or under `#defs`).

I've added a task to return to this.

Reply:
> We will definitely support array fields, and thus should validate constraints in them.

@CiaranMn to be clear would we support an array of objects?

Reply:
Correct, we would not support an array of nested objects unless we were supporting nested objects generally.